plugins/com.google.cloud.tools.eclipse.appengine.ui/src/com/google/cloud/tools/eclipse/appengine/ui/AppEngineLibrariesSelectorGroup.java

```diff
@@ -107,55 +104,205 @@ private static String getLibraryName(Library library) {
-      Library library = (Library) button.getData();
-      if (library.getId().equals(libraryId)) {
-        return button;
+  /**
+   * Tracks if the checkbox has been explicitly clicked by the user.
+   */
+  private static final class ManualSelectionTracker implements SelectionListener {
+    @Override
+    public void widgetSelected(SelectionEvent e) {
+      setManualSelection(e);
+    }
+
+    @Override
+    public void widgetDefaultSelected(SelectionEvent e) {
+      setManualSelection(e);
+    }
+
+    private void setManualSelection(SelectionEvent e) {
+      Button source = (Button)e.getSource();
+      if (e.getSource() instanceof Button && (source.getStyle() & SWT.CHECK) != 0) {
+        Button button = source;
+        button.setData(BUTTON_MANUAL_SELECTION_KEY, button.getSelection() ? new Object() : null);
       }
     }
-    return null;
   }
 
-  private Realm getDisplayRealm() {
-    return DisplayRealm.getRealm(parentDialog.getControl().getDisplay()); 
+  /**
+   * Returns a computed value based on whether the associated library is in a list or not. 
+   */
+  private static final class DependentLibrarySelected extends ComputedValue {
+    private String libraryId;
+    private IObservableList libraries;
+    private Getter<Boolean> condition;
+    private boolean selectedResult;
+
+    /**
+     * @param libraries the list of libraries to search
+     * @param libraryId the id of the library to be searched for
+     * @param foundResult return value if the library is found
+     */
+    private DependentLibrarySelected(Realm realm,
+                                     IObservableList libraries,
+                                     String libraryId,
+                                     final boolean foundResult) {
+      this(realm, libraries, libraryId, foundResult, new Getter<Boolean>(){
+        @Override
+        public Boolean get() {
+          return !foundResult;
+        }});
+    }
+
+    /**
+     * @param libraries the list of libraries to search
+     * @param libraryId the id of the library to be searched for
+     * @param foundResult return value if the library is selected
+     * @param condition if the library is not found in the list, return the result of <code>condition.get()</code>
+     */
+    private DependentLibrarySelected(Realm realm,
+                                     IObservableList libraries,
+                                     String libraryId,
+                                     boolean foundResult,
+                                     Getter<Boolean> condition) {
+      super(realm);
+      Preconditions.checkNotNull(libraries);
+      Preconditions.checkNotNull(libraryId);
+      Preconditions.checkNotNull(condition);
+      this.selectedResult = foundResult;
+      this.libraries = libraries;
+      this.libraryId = libraryId;
+      this.condition = condition;
+    }
+
+    @Override
+    protected Object calculate() {
+      for (int i = 0; i < libraries.size(); i++) {
```

Comment:
Enhanced `for`. `for (Object object : libraries)` will work.

I wish we were at Java 8.